File: frontend/src/engine/concrete.ts
// Concrete & RCC volumes and formwork. Ported from concrete.py. Units: m3 / m2.
import { Member } from "./members";
import { Quantity, mmToM, qty, step, fmt0 } from "./units";

const CLAUSE = "IS 1200 Part 2 / IS 456";

const grade = (m: Member) => m.concrete_grade ?? "M25";

export function column(m: Member): Quantity[] {
  const b = mmToM(m.b_mm), D = mmToM(m.D_mm), H = mmToM(m.height_mm);
  const n = m.count;
  const vol = b * D * H * n;
  const shutter = 2 * (b + D) * H * n;
  const label = `${grade(m)} RCC column ${m.label}`.trim();
  return [
    qty({
      category: "concrete", description: `${label} (${fmt0(m.b_mm)}x${fmt0(m.D_mm)} mm)`,
      unit: "m3", value: vol, nos: n, length_m: b, breadth_m: D, depth_m: H,
      audit: [step("concrete.column.volume", "b*D*H*count",
        { b_m: b, D_m: D, H_m: H, count: n }, vol, CLAUSE)],
      extra: { grade: grade(m) },
    }),
    qty({
      category: "formwork", description: `Formwork to column ${m.label}`.trim(),
      unit: "m2", value: shutter, nos: n,
      audit: [step("concrete.column.formwork", "2*(b+D)*H*count",
        { b_m: b, D_m: D, H_m: H, count: n }, shutter, CLAUSE)],
    }),
  ];
}

export function beam(m: Member): Quantity[] {
  const b = mmToM(m.b_mm), d = mmToM(m.depth_mm), L = mmToM(m.clear_span_mm);
  const n = m.count;
  const vol = b * d * L * n;
  const shutter = (2 * d + b) * L * n;
  const label = `${grade(m)} RCC beam ${m.label}`.trim();
  return [
    qty({
      category: "concrete", description: `${label} (${fmt0(m.b_mm)}x${fmt0(m.depth_mm)} mm)`,
      unit: "m3", value: vol, nos: n, length_m: L, breadth_m: b, depth_m: d,
      audit: [step("concrete.beam.volume", "b*d*L*count",
        { b_m: b, d_m: d, L_m: L, count: n }, vol, CLAUSE)],
      extra: { grade: grade(m) },
    }),
    qty({
      category: "formwork", description: `Formwork to beam ${m.label}`.trim(),
      unit: "m2", value: shutter, nos: n,
      audit: [step("concrete.beam.formwork", "(2*d+b)*L*count",
        { b_m: b, d_m: d, L_m: L, count: n }, shutter, CLAUSE)],
    }),
  ];
}

export function footing(m: Member): Quantity[] {
  const L = mmToM(m.length_mm), B = mmToM(m.breadth_mm), D = mmToM(m.depth_mm);
  const n = m.count;
  const vol = L * B * D * n;
  const shutter = 2 * (L + B) * D * n;   // vertical side shuttering of the pad
  const label = `${grade(m)} RCC footing ${m.label}`.trim();
  return [
    qty({
      category: "concrete", description: `${label} (${fmt0(m.length_mm)}x${fmt0(m.breadth_mm)} mm)`,
      unit: "m3", value: vol, nos: n, length_m: L, breadth_m: B, depth_m: D,
      audit: [step("concrete.footing.volume", "L*B*D*count",
        { L_m: L, B_m: B, D_m: D, count: n }, vol, CLAUSE)],
      extra: { grade: grade(m) },
    }),
    qty({
      category: "formwork", description: `Formwork to footing ${m.label}`.trim(),
      unit: "m2", value: shutter, nos: n,
      audit: [step("concrete.footing.formwork", "2*(L+B)*D*count",
        { L_m: L, B_m: B, D_m: D, count: n }, shutter, CLAUSE)],
    }),
  ];
}

export function slab(m: Member): Quantity[] {
  const L = mmToM(m.length_mm), B = mmToM(m.breadth_mm), t = mmToM(m.thickness_mm);
  const n = m.count;
  const gross_area = L * B;
  const net_area = Math.max(gross_area - (m.opening_area_m2 ?? 0), 0.0);
  const vol = net_area * t * n;
  const shutter = net_area * n;
  const label = `${grade(m)} RCC slab ${m.label}`.trim();
  return [
    qty({
      category: "concrete", description: `${label} (${fmt0(m.thickness_mm)} mm thick)`,
      unit: "m3", value: vol, nos: n, length_m: L, breadth_m: B, depth_m: t,
      audit: [step("concrete.slab.volume", "(L*B - openings)*t*count",
        { L_m: L, B_m: B, openings_m2: m.opening_area_m2 ?? 0, t_m: t, count: n }, vol, CLAUSE)],
      extra: { grade: grade(m) },
    }),
    qty({
      category: "formwork", description: `Formwork (soffit) to slab ${m.label}`.trim(),
      unit: "m2", value: shutter, nos: n,
      audit: [step("concrete.slab.formwork", "(L*B - openings)*count",
        { net_area_m2: net_area, count: n }, shutter, CLAUSE)],
    }),
  ];
}

export function rcc_wall(m: Member): Quantity[] {
  const L = mmToM(m.length_mm), H = mmToM(m.height_mm), t = mmToM(m.thickness_mm);
  const n = m.count;
  const vol = L * t * H * n;
  const shutter = 2 * L * H * n;
  const label = `${grade(m)} RCC wall ${m.label}`.trim();
  return [
    qty({
      category: "concrete", description: `${label} (${fmt0(m.thickness_mm)} mm thick)`,
      unit: "m3", value: vol, nos: n, length_m: L, breadth_m: t, depth_m: H,
      audit: [step("concrete.wall.volume", "L*t*H*count",
        { L_m: L, t_m: t, H_m: H, count: n }, vol, CLAUSE)],
      extra: { grade: grade(m) },
    }),
    qty({
      category: "formwork", description: `Formwork to wall ${m.label}`.trim(),
      unit: "m2", value: shutter, nos: n,
      audit: [step("concrete.wall.formwork", "2*L*H*count",
        { L_m: L, H_m: H, count: n }, shutter, CLAUSE)],
    }),
  ];
}

export function pcc(m: Member): Quantity[] {
  const L = mmToM(m.length_mm), B = mmToM(m.breadth_mm), t = mmToM(m.thickness_mm);
  const n = m.count;
  const vol = L * B * t * n;
  return [
    qty({
      category: "concrete", description: `PCC / lean concrete ${m.label} (${fmt0(m.thickness_mm)} mm)`.trim(),
      unit: "m3", value: vol, nos: n, length_m: L, breadth_m: B, depth_m: t,
      audit: [step("concrete.pcc.volume", "L*B*t*count",
        { L_m: L, B_m: B, t_m: t, count: n }, vol, CLAUSE)],
      extra: { grade: grade(m) },
    }),
  ];
}

File: backend/app/engine/concrete.py
"""Concrete & RCC volumes and formwork (shuttering). Units: m3 / m2.

See project.md section 6. All inputs in mm, converted to m here.
"""
from __future__ import annotations

from .units import FormulaStep, Quantity, mm_to_m

CLAUSE = "IS 1200 Part 2 / IS 456"


def _grade_desc(member) -> str:
    return getattr(member, "concrete_grade", "M25")


def column(m) -> list[Quantity]:
    b, D, H = mm_to_m(m.b_mm), mm_to_m(m.D_mm), mm_to_m(m.height_mm)
    n = m.count
    vol = b * D * H * n
    shutter = 2 * (b + D) * H * n
    label = f"{_grade_desc(m)} RCC column {m.label}".strip()
    return [
        Quantity(
            category="concrete", description=f"{label} ({m.b_mm:.0f}x{m.D_mm:.0f} mm)",
            unit="m3", value=vol, nos=n, length_m=b, breadth_m=D, depth_m=H,
            audit=[FormulaStep("concrete.column.volume", "b*D*H*count",
                               {"b_m": b, "D_m": D, "H_m": H, "count": n}, vol, CLAUSE)],
            extra={"grade": _grade_desc(m)},
        ),
        Quantity(
            category="formwork", description=f"Formwork to column {m.label}".strip(),
            unit="m2", value=shutter, nos=n,
            audit=[FormulaStep("concrete.column.formwork", "2*(b+D)*H*count",
                               {"b_m": b, "D_m": D, "H_m": H, "count": n}, shutter, CLAUSE)],
        ),
    ]


def beam(m) -> list[Quantity]:
    b, d, L = mm_to_m(m.b_mm), mm_to_m(m.depth_mm), mm_to_m(m.clear_span_mm)
    n = m.count
    vol = b * d * L * n
    shutter = (2 * d + b) * L * n          # two sides + soffit
    label = f"{_grade_desc(m)} RCC beam {m.label}".strip()
    return [
        Quantity(
            category="concrete", description=f"{label} ({m.b_mm:.0f}x{m.depth_mm:.0f} mm)",
            unit="m3", value=vol, nos=n, length_m=L, breadth_m=b, depth_m=d,
            audit=[FormulaStep("concrete.beam.volume", "b*d*L*count",
                               {"b_m": b, "d_m": d, "L_m": L, "count": n}, vol, CLAUSE)],
            extra={"grade": _grade_desc(m)},
        ),
        Quantity(
            category="formwork", description=f"Formwork to beam {m.label}".strip(),
            unit="m2", value=shutter, nos=n,
            audit=[FormulaStep("concrete.beam.formwork", "(2*d+b)*L*count",
                               {"b_m": b, "d_m": d, "L_m": L, "count": n}, shutter, CLAUSE)],
        ),
    ]


def footing(m) -> list[Quantity]:
    L, B, D = mm_to_m(m.length_mm), mm_to_m(m.breadth_mm), mm_to_m(m.depth_mm)
    n = m.count
    vol = L * B * D * n
    shutter = 2 * (L + B) * D * n          # vertical side shuttering of the pad
    label = f"{_grade_desc(m)} RCC footing {m.label}".strip()
    return [
        Quantity(
            category="concrete", description=f"{label} ({m.length_mm:.0f}x{m.breadth_mm:.0f} mm)",
            unit="m3", value=vol, nos=n, length_m=L, breadth_m=B, depth_m=D,
            audit=[FormulaStep("concrete.footing.volume", "L*B*D*count",
                               {"L_m": L, "B_m": B, "D_m": D, "count": n}, vol, CLAUSE)],
            extra={"grade": _grade_desc(m)},
        ),
        Quantity(
            category="formwork", description=f"Formwork to footing {m.label}".strip(),
            unit="m2", value=shutter, nos=n,
            audit=[FormulaStep("concrete.footing.formwork", "2*(L+B)*D*count",
                               {"L_m": L, "B_m": B, "D_m": D, "count": n}, shutter, CLAUSE)],
        ),
    ]


def slab(m) -> list[Quantity]:
    L, B, t = mm_to_m(m.length_mm), mm_to_m(m.breadth_mm), mm_to_m(m.thickness_mm)
    n = m.count
    gross_area = L * B
    net_area = max(gross_area - m.opening_area_m2, 0.0)
    vol = net_area * t * n
    shutter = net_area * n
    label = f"{_grade_desc(m)} RCC slab {m.label}".strip()
    return [
        Quantity(
            category="concrete", description=f"{label} ({m.thickness_mm:.0f} mm thick)",
            unit="m3", value=vol, nos=n, length_m=L, breadth_m=B, depth_m=t,
            audit=[FormulaStep("concrete.slab.volume", "(L*B - openings)*t*count",
                               {"L_m": L, "B_m": B, "openings_m2": m.opening_area_m2,
                                "t_m": t, "count": n}, vol, CLAUSE)],
            extra={"grade": _grade_desc(m)},
        ),
        Quantity(
            category="formwork", description=f"Formwork (soffit) to slab {m.label}".strip(),
            unit="m2", value=shutter, nos=n,
            audit=[FormulaStep("concrete.slab.formwork", "(L*B - openings)*count",
                               {"net_area_m2": net_area, "count": n}, shutter, CLAUSE)],
        ),
    ]


def rcc_wall(m) -> list[Quantity]:
    L, H, t = mm_to_m(m.length_mm), mm_to_m(m.height_mm), mm_to_m(m.thickness_mm)
    n = m.count
    vol = L * t * H * n
    shutter = 2 * L * H * n
    label = f"{_grade_desc(m)} RCC wall {m.label}".strip()
    return [
        Quantity(
            category="concrete", description=f"{label} ({m.thickness_mm:.0f} mm thick)",
            unit="m3", value=vol, nos=n, length_m=L, breadth_m=t, depth_m=H,
            audit=[FormulaStep("concrete.wall.volume", "L*t*H*count",
                               {"L_m": L, "t_m": t, "H_m": H, "count": n}, vol, CLAUSE)],
            extra={"grade": _grade_desc(m)},
        ),
        Quantity(
            category="formwork", description=f"Formwork to wall {m.label}".strip(),
            unit="m2", value=shutter, nos=n,
            audit=[FormulaStep("concrete.wall.formwork", "2*L*H*count",
                               {"L_m": L, "H_m": H, "count": n}, shutter, CLAUSE)],
        ),
    ]


def pcc(m) -> list[Quantity]:
    L, B, t = mm_to_m(m.length_mm), mm_to_m(m.breadth_mm), mm_to_m(m.thickness_mm)
    n = m.count
    vol = L * B * t * n
    return [
        Quantity(
            category="concrete", description=f"PCC / lean concrete {m.label} ({m.thickness_mm:.0f} mm)".strip(),
            unit="m3", value=vol, nos=n, length_m=L, breadth_m=B, depth_m=t,
            audit=[FormulaStep("concrete.pcc.volume", "L*B*t*count",
                               {"L_m": L, "B_m": B, "t_m": t, "count": n}, vol, CLAUSE)],
            extra={"grade": _grade_desc(m)},
        ),
    ]
